clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+# 
+# You will use the [harmonized Sentinal/Landsat multispectral dataset](https://lpdaac.usgs.gov/documents/1698/HLS_User_Guide_V2.pdf). You can access the data with an [Earthdata account](https://www.earthdata.nasa.gov/learn/get-started) by installing the [`earthaccess` library from NSIDC](https://github.com/nsidc/earthaccess):
+# 
+# ```bash
+# conda install -c conda-forge -y earthaccess
+# ```
+
+# %% [markdown]
+# ## STEP 1: SET UP
+
+# %% [markdown]
+# YOUR TASK:
+#   1. Import all libraries you will need for this analysis
+#   2. Configure GDAL parameters to help avoid connection errors:
+#      ```python
+#      os.environ["GDAL_HTTP_MAX_RETRY"] = "5"
+#      os.environ["GDAL_HTTP_RETRY_DELAY"] = "1"
+#      ```
+
+# %%
+import glob
+import itertools
+import json
+import logging
+import os
+import shutil
+import warnings
+from datetime import datetime
+
+import cartopy.crs as ccrs
+import earthpy as et
+import earthpy.plot as ep
+import earthaccess
+import geopandas as gpd
+import geoviews as gv
+import holoviews as hv
+import hvplot
+import hvplot.pandas
+import hvplot.xarray
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pystac_client
+import rasterio as rio
+import requests
+import rioxarray as rxr
+import rioxarray.merge as rxrmerge
+import xarray as xr
+from bokeh.util import logconfig
+from dotenv import load_dotenv
+from shapely.geometry import Polygon
+from sklearn.cluster import KMeans
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import StandardScaler
+
+# Supress warnings
+logconfig.basicConfig(level=logging.ERROR)
+warnings.filterwarnings('ignore')
+
+"""
+Loading these environment variables:
+    EARTHDATA_USERNAME
+    EARTHDATA_PASSWORD
+    GDAL_HTTP_MAX_RETRY=5
+    GDAL_HTTP_RETRY_DELAY=1
+    GDAL_HTTP_COOKIEFILE='~/cookies.txt'
+    GDAL_HTTP_COOKIEJAR='~/cookies.txt'
+    GDAL_DISABLE_READDIR_ON_OPEN='EMPTY_DIR'
+    CPL_VSIL_CURL_ALLOWED_EXTENSIONS='TIF'
+    GDAL_HTTP_UNSAFESSL='YES'
+"""
+load_dotenv() # Load .env file
+
+data_dir = os.path.join(et.io.HOME, et.io.DATA_NAME)
+cache_dir = os.path.join(data_dir, 'clustering-kmeans')
+
+if not os.path.exists(cache_dir):
+    os.makedirs(cache_dir)
+
+# %%
+username = os.getenv('EARTHDATA_USERNAME')
+password = os.getenv('EARTHDATA_PASSWORD')
+
+# earthaccess.login doesn't seem to be creating this file
+netrc_fpath = os.path.expanduser("~/.netrc")
+if not netrc_fpath:
+    with open(netrc_fpath, 'w') as file:
+        file.write(f"machine urs.earthdata.nasa.gov\nlogin {username}\npassword {password}\n")
```

Comment:
Interesting...I've not had this problem with earthaccess, but this is a nice solution. Skye and I did encounter an issue where on Windows sometimes the file is _netrc...so maybe that's related?

I like to split my strings at the new line so it's easier to read.